projects/transformers/trainer_mixins/deepspeed.py

```diff
@@ -0,0 +1,129 @@
+# ----------------------------------------------------------------------
+# Numenta Platform for Intelligent Computing (NuPIC)
+# Copyright (C) 2021, Numenta, Inc.  Unless you have an agreement
+# with Numenta, Inc., for a separate license for this software code, the
+# following terms and conditions apply:
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Affero Public License for more details.
+#
+# You should have received a copy of the GNU Affero Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+from collections import defaultdict
+
+import torch
+from deepspeed import replace_transformer_layer
+from deepspeed.module_inject import HFBertLayerPolicy
+from transformers import BertLayer
+
+from nupic.torch.modules import SparseWeights
+from projects.transformers.models.sparse_deepspeed_transformer import (
+    SparseDeepSpeedTransformerLayer,
+)
+
+__all__ = ["DeepspeedTransformerLayerMixin", "replace_sparse_transformer_layer"]
+
+
+def replace_sparse_transformer_layer(model, config=None, fp16=True, training=True):
+
+    # Store sparse mask and properties before deepspeed conversion
+    sparsity_per_layer = defaultdict(dict)
+    for i, bert_layer in enumerate(model.encoder.layer):
+        # Only supports huggingface BertLayer
+        assert isinstance(bert_layer, BertLayer)
+
+        # Check if model has sparse bert layers
+        query = bert_layer.attention.self.query
+        if isinstance(query, SparseWeights):
+            # Assume same sparsity for all QKV layers
```

Comment:
We have functionality to set a different sparsity per layer, though I've only used it in a a couple experiments. It could be good to specify this assumption in the docstring of the mixin.

Reply:
That said, you should definitely keep this assumption. Most likely, all of our experiments will abide by it.

Reply:
Done beb8eb7bcd8fb8a37c3e4fc1b111d4b250cd766f